c10/util/numa.cpp

```diff
@@ -0,0 +1,132 @@
+#include "c10/util/numa.h"
+
+C10_DEFINE_bool(caffe2_cpu_numa_enabled, false, "Use NUMA whenever possible.");
+
+#if defined(__linux__) && !defined(C10_DISABLE_NUMA) && C10_MOBILE == 0
+#include <numa.h>
+#include <numaif.h>
+#define C10_NUMA_ENABLED
+#endif
+
+namespace c10 {
+
+#ifdef C10_NUMA_ENABLED
+bool IsNUMAEnabled() {
+  return FLAGS_caffe2_cpu_numa_enabled && numa_available() >= 0;
+}
+
+void NUMABind(int numa_node_id) {
+  if (numa_node_id < 0) {
+    return;
+  }
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
+    return;
+  }
+
+  CAFFE_ENFORCE(
+      numa_node_id <= numa_max_node(),
+      "NUMA node id " + c10::to_string(numa_node_id) + " is unavailable");
+
+  auto bm = numa_allocate_nodemask();
+  numa_bitmask_clearall(bm);
+  numa_bitmask_setbit(bm, numa_node_id);
+  numa_bind(bm);
+  numa_bitmask_free(bm);
+}
+
+int GetNUMANode(const void* ptr) {
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
+    return -1;
+  }
+  CAFFE_ENFORCE(ptr);
+
+  int numa_node = -1;
+  CAFFE_ENFORCE(
+      get_mempolicy(
+          &numa_node, NULL, 0, (void*)ptr, MPOL_F_NODE | MPOL_F_ADDR) == 0,
+      "Unable to get memory policy");
+  return numa_node;
+}
+
+int GetNumNUMANodes() {
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
+    return -1;
+  }
+
+  return numa_num_configured_nodes();
+}
+
+void NUMAMove(void* ptr, size_t size, int numa_node_id) {
+  if (numa_node_id < 0) {
+    return;
+  }
+  if (!IsNUMAEnabled()) {
+    VLOG(1) << "NUMA is not enabled";
+    return;
+  }
+  CAFFE_ENFORCE(ptr);
+
+  size_t page_start_ptr = (((size_t)ptr) & ~(getpagesize() - 1));
```

Comment:
1. This is using a C-style cast, we should not use C-style casts.
2. The standard doesn't guarantee that casting a pointer to `size_t` to be well defined. Use `uintptr_t` instead.